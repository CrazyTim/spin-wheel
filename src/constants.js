/**
 * Adjustment when drawing canvas arcs, because it draws from 90° instead of 0°.
 */
export const arcAdjust = -90;

export const baseCanvasSize = 500; // 500 seemed to be a good value for this.

/**
 * The period in milliseconds that we record drag events.
 * Used to calculate how fast the wheel should spin after the drag ends.
 * For example, if the wheel was dragged 20 degrees over the last 250ms,
 * then it should continue rotating at a speed of 20 degrees every 250ms after the drag ends.
 */
export const dragCapturePeriod = 250;

/**
 * Text alignment enum.
 */
export const AlignText = Object.freeze({
  left: 'left',
  right: 'right',
  center: 'center',
});

/**
 * Wheel property defaults.
 */
export const Defaults = Object.freeze({
  wheel: {
    borderColor: '#000',
    borderWidth: 1,
    debug: false,
    image: null,
    isInteractive: true,
    itemBackgroundColors: ['#fff'],
    itemLabelAlign: AlignText.right,
    itemLabelBaselineOffset: 0,
    itemLabelColors: ['#000'],
    itemLabelFont: 'sans-serif',
    itemLabelFontSizeMax: baseCanvasSize,
    itemLabelRadius: 0.85,
    itemLabelRadiusMax: 0.2,
    itemLabelRotation: 0,
    items: [],
    lineColor: '#000',
    lineWidth: 1,
    pixelRatio: 0,
    radius: 0.95,
    rotation: 0,
    rotationResistance: -35,
    rotationSpeed: 0,
    rotationSpeedMax: 250,
    offset: {w: 0, h: 0},
    onCurrentIndexChange: null,
    onRest: null,
    onSpin: null,
    overlayImage: null,
    pointerAngle: 0,
  },
  item: {
    backgroundColor: null, // A value of null causes this to fall back to `Wheel.itemBackgroundColors`
    image: null,
    imageRadius: 0.5,
    imageRotation: 0,
    imageScale: 1,
    label: '',
    labelColor: null, // A value of null causes this to fall back to `Wheel.itemLabelColors`
    labelFont: null, // A value of null causes this to fall back to `Wheel.itemLabelFont`
    weight: 1,
  },
});

export const Debugging = Object.freeze({
  pointerLineColor: '#ff00ff',
  labelOutlineColor: '#ff00ff',
  labelRadiusColor: '#00ff00',
  dragEventHue: 200,
});
