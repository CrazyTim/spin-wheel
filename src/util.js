/**
 * Get a random integer between `min` (inclusive) and `max` (exclusive).
 */
export function getRandomInt(min = 0, max = 0) {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min)) + min;
}

/**
 * Get a random number between `min` (inclusive) and `max` (inclusive).
 * Control the number of decimal places with `round`.
 */
 export function getRandomFloat(min = 0, max = 0, round = 14) {
  return parseFloat((Math.random() * (max - min) + min).toFixed(round));
}

/**
 * Convert degrees to radians.
 */
export function degRad(degrees = 0) {
  return degrees * Math.PI / 180;
}

/**
 * Return true if `angle` is between `arcStart` (inclusive) and `arcEnd` (exclusive).
 * Example: `(0, 359, 1) == true`
 * Example: `(0, 1, 2) == false`
 */
export function isAngleBetween(angle, arcStart, arcEnd) {
  if (arcStart < arcEnd)
   return arcStart <= angle && angle < arcEnd;
   return arcStart <= angle || angle < arcEnd;
}

/**
 * Average the values in `array`.
 * Only operate on truthy values.
 * Truthy values that are not Numbers count as 1.
 * Return 0 if the array is empty.
 */
export function aveArray(array = []) {
  let sum = 0;
  for (const val of array) {
    if (val) sum += ((typeof val === 'number') ? val : 1);
  }
  return sum / array.length || 0;
}

/**
 * Calculate the largest font size that `text` can have without exceeding `maxWidth`.
 * Won't work unless `fontFamily` has been loaded.
 */
export function getFontSizeToFit(text, fontFamily, maxWidth, canvasContext) {
  canvasContext.save();
  canvasContext.font = `1px ${fontFamily}`;
  const w = canvasContext.measureText(text).width;
  canvasContext.restore();
  return maxWidth / w;
}

/**
 * Return true if the given point is inside a circle.
 * cx, cy is circle center.
 * radius is circle radius.
 */
export function isPointInCircle(point = {x: 0, y: 0}, cx, cy, radius) {
  const distanceSquared = ((point.x - cx) ** 2) + ((point.y - cy) ** 2);
  return distanceSquared <= (radius ** 2);
}

/**
 * Translate the given point from the viewport's coordinate space to the element's coordinate space.
 */
export function translateXYToElement(point = {x: 0, y: 0}, element = {}, devicePixelRatio = 1) {
  const rect = element.getBoundingClientRect();
  return {
    x: (point.x - rect.left) * devicePixelRatio,
    y: (point.y - rect.top) * devicePixelRatio,
  };
}

export function getMouseButtonsPressed(event = {}) {
  return [1, 2, 4, 8, 16].filter(i => event.buttons & i);
}

/**
 * Source: https://stackoverflow.com/a/47653643/737393
 */
export function getAngle(originX, originY, targetX, targetY) {
    const dx = originX - targetX;
    const dy = originY - targetY;

    let theta = Math.atan2(-dy, -dx);
    theta *= 180 / Math.PI;
    if (theta < 0) theta += 360;

    return theta;
}

/**
 * Return the distance between two points.
 */
export function getDistanceBetweenPoints(point1 = {x: 0, y: 0}, point2 = {x: 0, y: 0}) {
  return Math.hypot(point2.x - point1.x, point2.y - point1.y);
}

/**
 * Add two angles together.
 * Return a positive number between 0 and 359.9999.
 */
export function addAngle(a = 0, b = 0) {
  const sum = a + b;
  let result;

  if (sum > 0) {
    result = sum % 360;
  } else {
    result = 360 + (sum % 360);
  }
  if (result === 360) result = 0;

  return result;
}

/**
 * Return the shortest difference between two angles.
 * Only accept angles between 0 and 360.
 */
export function diffAngle(a = 0, b = 0) {
  const offsetFrom180 = 180 - b;
  const aWithOffset = addAngle(a, offsetFrom180);
  return 180 - aWithOffset;
}

export function isObject(v) {
  return typeof v === 'object' && !Array.isArray(v) && v !== null;
}

export function easeSinOut(n) {
  return Math.sin((n * Math.PI) / 2);
}