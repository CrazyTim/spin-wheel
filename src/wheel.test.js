import {beforeAll, test, expect} from '@jest/globals';
import * as util from './util.js';
import {Defaults} from './constants.js';
import {Wheel} from './wheel.js';
import setup from '../scripts/test-setup.js';
import {getInstanceProperties} from '../scripts/util.js';

beforeAll(() => {
  setup.useWheel();
});

test('Each Wheel property has a coresponding default value', () => {

  const wheelGetters = getInstanceProperties(window.wheel).getters;

  for (const i of wheelGetters) {
    expect(Defaults.wheel[i]).not.toBe(undefined);
  }

});

test('Each Wheel property is given a Default value when the Wheel is initalised', () => {

  const wheelGetters = getInstanceProperties(window.wheel).getters;

  for (const i of wheelGetters) {
    expect(window.wheel[i]).toEqual(Defaults.wheel[i]);
  }

});

test ('getItemAngles() works', () => {

  setup.addBlankItems(4);
  window.wheel.rotation = 90;

  const angles = window.wheel.getItemAngles();

  expect(angles[0].start).toBe(0); // First start angle should be 0.
  expect(angles[3].end).toBe(angles[0].start + 360); // Last angle.end must be the same as first angle.start.

  expect(angles[0].end).toBe(90);
  expect(angles[1].start).toBe(90);
  expect(angles[1].end).toBe(180);

});

test ('getItemAngles() works when weighted', () => {

  window.wheel.items = [
    {weight: 2},
    {weight: 1},
    {weight: 1},
  ];

  const angles = window.wheel.getItemAngles();

  expect(angles[0].start).toBe(0);
  expect(angles[0].end).toBe(180);
  expect(angles[1].start).toBe(180);
  expect(angles[1].end).toBe(270);

});
