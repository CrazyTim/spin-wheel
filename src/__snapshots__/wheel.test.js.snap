// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wheel default state is correct 1`] = `
Wheel {
  "_actualRadius": 237.5,
  "_borderColor": "#000",
  "_borderWidth": 1,
  "_canvasContainer": <div>
    <canvas
      height="500"
      style="width: 500px; height: 500px;"
      width="500"
    />
  </div>,
  "_center": Object {
    "x": 250,
    "y": 250,
  },
  "_context": CanvasRenderingContext2D {
    "canvas": <canvas
      height="500"
      style="width: 500px; height: 500px;"
      width="500"
    />,
    "createPattern": [Function],
    "drawImage": [Function],
  },
  "_currentIndex": -1,
  "_debug": false,
  "_frameRequestId": 1,
  "_handler_onMouseDown": [Function],
  "_handler_onMouseMoveRefreshCursor": [Function],
  "_handler_onPointerDown": [Function],
  "_handler_onPointerMoveRefreshCursor": [Function],
  "_handler_onResize": [Function],
  "_handler_onTouchStart": [Function],
  "_image": null,
  "_isInitialising": true,
  "_isInteractive": true,
  "_itemBackgroundColors": Array [
    "#fff",
  ],
  "_itemLabelAlign": "right",
  "_itemLabelBaselineOffset": 0,
  "_itemLabelColors": Array [
    "#000",
  ],
  "_itemLabelFont": "sans-serif",
  "_itemLabelFontSizeMax": 500,
  "_itemLabelRadius": 0.85,
  "_itemLabelRadiusMax": 0.2,
  "_itemLabelRotation": 0,
  "_items": Array [],
  "_lineColor": "#000",
  "_lineWidth": 1,
  "_offset": Object {
    "h": 0,
    "w": 0,
  },
  "_onCurrentIndexChange": null,
  "_onRest": null,
  "_onSpin": null,
  "_overlayImage": null,
  "_pixelRatio": 0,
  "_pointerAngle": 0,
  "_radius": 0.95,
  "_rotation": 0,
  "_rotationDirection": 0,
  "_rotationResistance": -35,
  "_rotationSpeed": 0,
  "_rotationSpeedMax": 250,
  "_size": 500,
  "canvas": <canvas
    height="500"
    style="width: 500px; height: 500px;"
    width="500"
  />,
  "itemLabelFontSize": 500,
  "labelMaxWidth": 154.37499999999997,
}
`;
